### PR TITLE
#1864 - MSFAA Send File E2E Test Automation

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Created the method `getUploadedFile` to allow access to the supposed-to-be uploaded file.
- Configure tests debug to use the timezone as UTC, otherwise, the expected result will differ from the real ones.
- Create a specific factory (`saveMSFAATestInputData`) to generate the MSFAA records that could be used for all the MSFAA file-related tests.
- Centralized the `processDate` used for MSFAA records, MSFAA file name, and MSFAA sequence group.
- Added an "order by" on the `getPendingMSFAARequest` to ensure that the records are consistently retrieved in the same order and as a consequence, the file records will be generated using the same order also.
- Created the helpers `createE2EDataSources` to be used specifically for E2E tests and reduce the amount of code needed when saving to DB multiple entities. The intention is to be used as below. Open for discussions and opinions.
image
This is related to the idea shared on the DEVs chat on April 11th
image

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  // SIN validation.
+  const sinValidation = createFakeSINValidation({ student: fakeStudent });
+  sinValidation.sin = msfaaDataInput.sin;
+  // Student.
+  const student = await saveFakeStudent(db.dataSource, {
+    student: fakeStudent,
+    sinValidation,
+  });
+  // Application, offering, location.
+  const referenceApplication = await saveFakeApplication(db.dataSource, {});
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts:55]: do we need 2nd parameter here?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-helper.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import * as dayjs from "dayjs";
+
+/**
+ * Creates the sequence group name by the offering intensity.
+ * @param offeringIntensity offering intensity.
+ * @returns the sequence group name by the offering intensity.
+ */
+export function getMSFAASequenceGroupName(
+  offeringIntensity: OfferingIntensity,
+) {
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-helper.ts:18]: return type?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.models.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  institutionCode: "EFDG",
+  birthDate: "2000-01-01",
+  surname: "Other Doe",
+  givenName: "Jane",
+  gender: "unknown",
+  maritalStatus: RelationshipStatus.Single,
+  addressLine1: "Address Line 1",
+  addressLine2: undefined,
+  city: "Some city on United States",
+  provinceState: "BC",
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.models.ts:58]: mm, these are added by the students, right? I believe we don't have `provinceState` when the country is different right?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-helper.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * @returns processDate and the perspectives formatted date and time.
+ */
+export function getProcessDateFromMSFAARequestContent(
+  mockedMethod: jest.SpyInstance,
+): {
+  processDate: Date;
+  processDateFormatted: string;
+  processTimeFormatted: string;
+} {
+  // Method expected to be called once and receive the processDate parameter at index 3.
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-helper.ts:53]: 👍

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      const msfaaIDs = createdMSFAARecords.map((msfaa) => msfaa.id);
+      const msfaaUpdatedRecords = await db.msfaaNumber.find({
+        select: {
+          dateRequested: true,
+        },
+        where: {
+          id: In(msfaaIDs),
+        },
+      });
+      const allMSFAAsHaveDateRequested = msfaaUpdatedRecords.every((msfaa) =>
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts:141]: 👍

File: sources/packages/backend/libs/test-utils/src/data-source/e2e-data-source.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+} from "@sims/sims-db";
+import { DataSource, Repository } from "typeorm";
+
+/**
+ * Database helper to provide easy access to all repositories as needed.
+ * Intended to be used exclusively for E2E tests.
+ * @param dataSource Nestjs data source that provides access to all repositories.
+ * @returns database repositories.
+ */
+export function createE2EDataSources(dataSource: DataSource): E2EDataSources {
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/data-source/e2e-data-source.ts:56]: nice 👍

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts [hunk 7]
```diff
@@ -23,7 +23,7 @@ export class CreateTestingModuleResult {
   module: TestingModule;
   dataSource: DataSource;
   zbClient: ZBClient;
-  sshClientMock: Client;
+  sshClientMock: DeepMocked<Client>;
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:26]: for my info, what does `DeepMocked` do

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts:55]: Nope, my mistake.

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts]: Removed second parameter.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-helper.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-helper.ts]: Added return type.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.models.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.models.ts]: Set `provinceState` as `undefined`.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.models.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  msfaaNumber: string;
+  sin: string;
+  institutionCode: string;
+  birthDate: string;
+  surname: string;
+  givenName: string;
+  gender: string;
+  maritalStatus: RelationshipStatus;
+  addressLine1: string;
+  addressLine2: string;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.models.ts:16]: Looking at the file-details, few of these inputs could be optional right? May be that could become a scenario in future if needed

image

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  const newMSFAANumber = createFakeMSFAANumber({
+    student,
+    referenceApplication,
+  });
+  newMSFAANumber.msfaaNumber = msfaaDataInput.msfaaNumber;
+  newMSFAANumber.dateRequested = null;
+  newMSFAANumber.dateSigned = null;
+  newMSFAANumber.serviceProviderReceivedDate = null;
+  newMSFAANumber.cancelledDate = null;
+  newMSFAANumber.newIssuingProvince = null;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts:73]: IMO, I feel like this logic can be kept inside the factory `createFakeMSFAANumber` based on an optional parameter provided to the factory.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.models.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.models.ts:16]: They can be test scenarios but not necessarily E2E scenarios. The idea is to have some E2E and focus more on the unit tests for the queue consumers, as per my understanding, to compensate also the decision to have E2E running sequentially.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts:73]: It follows the idea of having the factory create the record and adjust the test as needed.
Keep in mind that these "nulls" are also set once inside a "factory".
Is the idea to create options like "pending", "canceled", or "signed" states?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        MSFAA_PART_TIME_RELATIONSHIP_OTHER,
+      ];
+      const createdMSFAARecords = await saveMSFAATestInputsData(
+        db,
+        msfaaInputData,
+      );
+      // Queued job.
+      const job = createMock<Job<void>>();
+      // Spy on the MSFAA content creation to intercept the process date and time used.
+      const createMSFAARequestContentMock =
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts:79]: Are we expecting more usage of `createMSFAARequestContentSpyOnMock()`?  I assume it should be for full time msfaa as well.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+      // Cancel any pending MSFAA.
+      await db.msfaaNumber.update(
+        { cancelledDate: IsNull() },
+        { cancelledDate: getISODateOnlyString(new Date()) },
+      );
+    });
+
+    it("Should generate an MSFAA part-time file and update the dateRequested when there are pending MSFAA records.", async () => {
+      // Arrange
+      const msfaaInputData: MSFAATestInputData[] = [
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts:67]: For my info, should we explicitly add the type here if we're setting the var?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+      // Assert
+      expect(createMSFAARequestContentMock).toHaveBeenCalledTimes(1);
+      const { processDate, processDateFormatted, processTimeFormatted } =
+        getProcessDateFromMSFAARequestContent(createMSFAARequestContentMock);
+      const uploadedFile = getUploadedFile(sftpClientMock);
+      expect(uploadedFile.remoteFilePath).toBe(
+        `MSFT-Request\\DPBC.EDU.MSFA.SENT.PT.${processDateFormatted}.001`,
+      );
+      // Assert process result.
+      expect(msfaaRequestResults).toHaveLength(1);
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts:94]: can't we directly assert like this ? 

```ts
expect(msfaaRequestResults).toBe([
        {
          generatedFile: uploadedFile.remoteFilePath,
          uploadedRecords: msfaaInputData.length,
          offeringIntensity: OfferingIntensity.partTime,
        },
      ]);
```

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts:67]: Added for an ode that was changed, no longer needed, changing it.

File: sources/packages/backend/package.json [hunk 13]
```diff
@@ -22,13 +22,13 @@
...
     "test:e2e:workers": "npm run migration:run && cross-env ENVIRONMENT=test jest --collect-coverage --verbose --config ./apps/workers/test/jest-e2e.json --forceExit",
     "test:e2e:workflow:local": "cross-env ENVIRONMENT=test jest --verbose --config ./workflow/test/jest-e2e.json --forceExit",
     "test:e2e:workflow": "npm run deploy:camunda:definitions && cross-env ENVIRONMENT=test jest --verbose --config ./workflow/test/jest-e2e.json --forceExit",
-    "test:e2e:queue-consumers": "npm run migration:run && cross-env ENVIRONMENT=test jest --collect-coverage --verbose --config ./apps/queue-consumers/test/jest-e2e.json --forceExit",
-    "test:e2e:queue-consumers:local": "cross-env ENVIRONMENT=test jest --verbose --config ./apps/queue-consumers/test/jest-e2e.json --forceExit",
+    "test:e2e:queue-consumers": "npm run migration:run && cross-env ENVIRONMENT=test jest --collect-coverage --verbose --config ./apps/queue-consumers/test/jest-e2e.json --runInBand --forceExit",
+    "test:e2e:queue-consumers:local": "cross-env ENVIRONMENT=test TZ=UTC jest --verbose --config ./apps/queue-consumers/test/jest-e2e.json --runInBand --forceExit",
```
Comment (dheepak-aot) [inline sources/packages/backend/package.json:31]: 👍

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts]: Changed.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+} from "@sims/test-utils";
+import { MSFAATestInputData } from "./msfaa-part-time-process-integration.scheduler.models";
+import { createFakeSINValidation } from "@sims/test-utils/factories/sin-validation";
+
+/**
+ * Save an MSFAA record providing all data that will be
+ * part of the MSFAA file generation.
+ * @param db data source helper.
+ * @param msfaa test input data.
+ * @returns a saved MSFAA record that used th input
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-factory.ts]: Fixed.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-integration.scheduler.e2e-spec.ts]: Nice catch, changed but it was required to use `toStrictEqual`.

**Review by andrepestana-aot — APPROVED** [review]
Great job! Looks good!

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts [hunk 15]
```diff
@@ -11,7 +11,7 @@ import {
   overrideImportsMetadata,
 } from "@sims/test-utils";
 import * as Client from "ssh2-sftp-client";
-import { createMock } from "@golevelup/ts-jest";
+import { DeepMocked, createMock } from "@golevelup/ts-jest";
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:14]: Not a part of the PR, I have question on line 45
Will this automatically mock the client.put which uploads the file?
`const sshClientMock = createMock<Client>();`

image

**Comment by dheepak-aot** [comment]
Thanks for setting up the E2E tests for file integration. Added few minor comments/questions.

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts (hunk 15 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:14]: Yes, `createMock<Client>();` is a helper to do a "deep" mock.

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the code change and the walk through. LGTM 👍

**Review by ann-aot — APPROVED** [review]
Good work @andrewsignori-aot 👍

File: sources/packages/backend/libs/integrations/src/esdc-integration/msfaa-integration/msfaa-request.processing.service.ts [hunk 16]
```diff
@@ -102,7 +104,7 @@ export class MSFAARequestProcessingService extends ESDCFileHandler {
           const msfaaNumberRepo = entityManager.getRepository(MSFAANumber);
           this.msfaaNumberService.updateRecordsInSentFile(
             msfaaRecordIds,
-            getUTCNow(),
+            processDate,
```
Comment (guru-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/msfaa-integration/msfaa-request.processing.service.ts:107]: 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

**Pull request merged**